communication/src/PlainUDPCommunication.cpp

```diff
@@ -321,8 +308,12 @@ class PlainUDPCommunication::PlainUdpImpl {
 #endif
     int mLen = 0;
     do {
-      mLen =
-          recvfrom(udpSockFd, bufferForIncomingMessages, maxMsgSize, 0, (sockaddr *)&fromAddress, &fromAddressLength);
+      mLen = recvfrom(udpSockFd,
+                      bufferForIncomingMessages,
+                      maxMsgSize,
+                      MSG_PEEK | MSG_DONTWAIT,
```

Comment:
This flag is resulting in a data race. Since, the flow is blocked here, the another thread trying to close the socket results in data race. With this change, the code will read the msg, if available. The code integrity is still maintained.

Reply:
The offline discussion:
With the flag `MSG_PEEK | MSG_DONTWAIT`, `recvfrom` does not block if there is no data in the socket and simply goes to `while(running)`. As a result, it might lead to a busy loop for connections with a low request rate.

Agreed on the following:
* Check if networking API is thread-safe in Linux
* If so, revert the code to original
* If no, find another way to unblock `recvfrom` when `Stop` is called.

As an option, we may create a pipe and wait in `select` for `udpSockFd` and the "reader" socket of the pipe. Once we want to cancel waiting on `udpSockFd`, the `Stop` thread can write to the "writer" socket of the pipe.

Reply:
use poll / epoll instead of select as behaviour of select is undefined if we are polling a fd which is greater than 1024.

Reply:
Changes done. `poll` used with timeout of 5 seconds.